### ISSUE TITLE
[Workaround] Temporally fix for Smartplay feature

### ISSUE THREAD

**Issue by psp2111-ADSLGATE** [nixgates/plugin.video.seren#923]
This is a Fix for Smartplay known BUG with TMDb Helper, using @inb4after Seren Fork and Modified UpNext addon by @MoojMidge special Thanks to them.

**Download**:
- [Seren fork](https://github.com/inb4after/plugin.video.seren/archive/refs/heads/master.zip) by inb4after.
- [UpNext](https://github.com/MoojMidge/service.upnext/releases) by MoojMidge.
- Latest TMDb Helper "currently v5.1.28"
- [Seren Player](https://github.com/mrgsi/tmdbh.players).

**How to Setup**:
Seren Settings → Playback:
- Check ✔️ Enable Smartplay and Pre-emptive scraping.
- change "minimum seconds left.." to 10s
- disable "playing next dialog" and "still watching dialog".
- press OK.
UpNext Settings → Expert:
- Check ✔️ Enable TMDb Helper integration. 
- Uncheck "exact match only" and "queue all next episodes"
- press OK
- ALSO YOU CAN CUSTOMISE "Behaviour" as you prefer .
Restart Kodi.

NOTES: 
- This is only a workaround by using Modified UpNext addon by @MoojMidge and making it take care of "Next episode popup" instead of Seren, so that mean UpNext could works, showup or interferen with other players not just Seren, for example: this workaround also work with Umbrella v6.626 Source select player, even so Umbrella does not natively support that.
- ONLY TESTED WITH Seren source select players.
- Pre-emptive scraping works half the time meaning: 1st ep works, 2nd not, 3rd works, 4th not... and so on.

**Comment by psp2111-ADSLGATE** [comment]
@cgwprs @eengert @welovewebs @unicornsandcandies

**Comment by MoojMidge** [comment]
> * Pre-emptive scraping works half the time meaning: 1st ep works, 2nd not, 3rd works, 4th not... and so on.

Do you have a debug log for this? It can probably be fixed.

**Comment by psp2111-ADSLGATE** [comment]
> > * Pre-emptive scraping works half the time meaning: 1st ep works, 2nd not, 3rd works, 4th not... and so on.
> 
> Do you have a debug log for this? It can probably be fixed.

I could make a quick debug log but do I need to enable any debug logging inside TMDb Helper or UpNext?

I am not 100% sure but I think it is a Bug in Seren not UpNext since SmartPlay almost behave the same way.

**Comment by MoojMidge** [comment]
Ah ok. I don't use Seren so I don't know what SmartPlay or Pre-emptive scraping actually do.

If it is something that can be fixed in UpNext then just a normal debug logging will be fine.

**Comment by psp2111-ADSLGATE** [comment]
> Ah ok. I don't use Seren so I don't know what SmartPlay or Pre-emptive scraping actually do.
> 
> If it is something that can be fixed in UpNext then just a normal debug logging will be fine.

Here is my log I played 4 episodes.
[kodi.log](https://github.com/nixgates/plugin.video.seren/files/12266722/kodi.log)
Start playing episodes at 2023-08-05 12:13:53.192 T:2716

About Smart Play as far as I understand it is just a build in UpNext inside Seren, with the following features: **Play Next** popup, **Still watching** popup, **Skip intro** and **Pre-emptive scraping**.
for Pre-emptive scraping: it is just to make Seren search for links saliently in the background while you watching the episode and after you finish watching and hit "Watch Now" for the next episode the results will show up immediately without waiting for Seren to scrape it.

Quick unrelated question: 
Can I make the previous episode "pause or stop" after pressing "Watch Now"? 
Becuase right now the previous will keep playing until I choose the next one!

**Comment by MoojMidge** [comment]
Thanks for the explanation of Smart Play and Pre-emptive scraping. I suspect I know what the problem is - the pre-emptive scraping processes the next playlist item and then runs the plugin script, which changes the next item from a plugin url to a scraped video url. After this is done, UpNext is unable to do anything because there is no plugin url that it can use.

In this case, you are right, it is a Seren issue. Ideally Seren should store the pre-scraped video url and use xbmcplugin.setResolvedUrl when the original plugin url is called, but there may be specific technical reasons why this is not done.

> Here is my log I played 4 episodes.

Sorry, should have been more clear. The log supplied is not a debug log, just a regular event log. I will need either a Kodi debug log: https://kodi.wiki/view/Log_file/Easy#Steps; or enable debug logging in UpNext: `UpNext > Settings > Expert > Troubleshooting > Log level > Debug`, otherwise UpNext won't log anything useful.

If you can get this, then I can see whether anything can be done in UpNext to workaround the problem.

> Can I make the previous episode "pause or stop" after pressing "Watch Now"?

It is easy enough to add this, but it may cause issues with the Kodi playback event order sequence, which is why I never tried to implement this.

**Comment by psp2111-ADSLGATE** [comment]
> Sorry, should have been more clear. The log supplied is not a debug log, just a regular event log. I will need either a Kodi debug log: https://kodi.wiki/view/Log_file/Easy#Steps; or enable debug logging in UpNext: UpNext > Settings > Expert > Troubleshooting > Log level > Debug, otherwise UpNext won't log anything useful.

no problem man here a new one with debug logging enabled.
[kodi.log](https://github.com/nixgates/plugin.video.seren/files/12267810/kodi.log)

>It is easy enough to add this, but it may cause issues with the Kodi playback event order sequence, which is why I never tried to implement this.

That a bummer, it will be nice to have and make the UpNext more polish.

**Comment by MoojMidge** [comment]
It looks like everything is actually working properly, except the TMDB Helper player file that you are using doesn't use pre-emptive scraping. You could try changing `action=getSources` to `action=preScrape` in the player file and see if it works.

**Comment by psp2111-ADSLGATE** [comment]
> You could try changing `action=getSources` to `action=preScrape` in the player file and see if it works.

Nope. that bricked the player, now only scraping and does not show the results window after finishing.

**Comment by MoojMidge** [comment]
🤷 

There probably should be a way to get it to work, but I don't use these plugins so I can't help.

**Comment by Needmorepixels** [comment]
I'm getting a the dependency on xbmc.python version 2.25 could not be satisfied when trying to install the modded up next. I'm running 20.2  and it was tested on 21.2 which I presume runs Python 3 so I'm not sure how to get this to work or how it worked for you guys. Appreciate any help as I'm really not a tech savvy guy!

**Comment by MoojMidge** [comment]
@Needmorepixels - the zip file with `matrix` in the filename is for Kodi 19+, you have downloaded a version for Kodi 18 which is why you are getting that error.
